src/smt/proof_manager.cpp

```diff
@@ -161,7 +161,7 @@ void PfManager::getAssertions(Assertions& as,
     assertions.push_back(*i);
   }
   NodeManager* nm = NodeManager::currentNM();
-  for (const std::pair<Node, smt::DefinedFunction>& dfn : df)
+  for (const std::pair<const Node, const smt::DefinedFunction>& dfn : df)
```

Comment:
This is purely about the type (for `dfn`) matching what is in `df`. `df` is a `context::CDHashMap<Node, smt::DefinedFunction, NodeHashFunction>` whose `value_type` ends up to be `std::pair<const Key, const Data>`. If you use anything else, the for loop needs to copy the whole thing.
(The mere fact that we discuss this makes my point for using `auto` more often ;-) )

Reply:
Would `auto` generate the type with the `const`? At some point I was trying to have auto generate a `const_iterator` rather than an `iterator` as a result of some `find` and apparently it was not possible force this. Is something like this not an issue here?

Reply:
The leftmost "const" is optional and depends on if the iterator can modify the data.  

The "const" before "const Node" should always be there since that is the value_type as Gereon mentions.

Reply:
I see, so in this case `const auto&` is fine. And are there cases where `const auto&` would be different from manually doing `const std::pair<const Key, const Data>&`?

Reply:
It depends on the container: for `CDHashMap` it will always be `std::pair<const Key, const Data>`. For `std::map` it would be `std::pair<const Key, Data>`, for `std::vector<T>` it would just be `T`.